tests/test_user_context.py

```diff
@@ -418,3 +419,697 @@ def test_decide_feature_rollout__send_flag_decision_false(self):
+                'rule_key': expected.rule_key,
+                'reasons': expected.reasons,
+                'decision_event_dispatched': False,
+                'variables': expected.variables,
+            },
+        )
+
+        # assert event count
+        self.assertEqual(0, mock_send_event.call_count)
+
+    def test_decide__option__exclude_variables(self):
+        opt_obj = optimizely.Optimizely(json.dumps(self.config_dict_with_features))
+        project_config = opt_obj.config_manager.get_config()
+
+        mock_experiment = project_config.get_experiment_from_key('test_experiment')
+        mock_variation = project_config.get_variation_from_id('test_experiment', '111129')
+
+        with mock.patch(
+                'optimizely.decision_service.DecisionService.get_variation_for_feature',
+                return_value=(decision_service.Decision(mock_experiment, mock_variation,
+                                                        enums.DecisionSources.FEATURE_TEST), []),
+        ), mock.patch(
+            'optimizely.notification_center.NotificationCenter.send_notifications'
+        ) as mock_broadcast_decision, mock.patch(
+            'optimizely.optimizely.Optimizely._send_impression_event'
+        ) as mock_send_event:
+
+            user_context = opt_obj.create_user_context('test_user', {'browser': 'chrome'})
+            actual = user_context.decide('test_feature_in_experiment', ['EXCLUDE_VARIABLES'])
+
+        expected_variables = {}
+
+        expected = OptimizelyDecision(
+            variation_key='variation',
+            rule_key='test_experiment',
+            enabled=True,
+            variables=expected_variables,
+            flag_key='test_feature_in_experiment',
+            user_context=user_context
+        )
+
+        self.compare_opt_decisions(expected, actual)
+
+        # assert notification
+        mock_broadcast_decision.assert_called_with(
+            enums.NotificationTypes.DECISION,
+            'flag',
+            'test_user',
+            {'browser': 'chrome'},
+            {
+                'flag_key': expected.flag_key,
+                'enabled': expected.enabled,
+                'variation_key': expected.variation_key,
+                'rule_key': expected.rule_key,
+                'reasons': expected.reasons,
+                'decision_event_dispatched': True,
+                'variables': expected.variables,
+            },
+        )
+
+        # assert event count
+        self.assertEqual(1, mock_send_event.call_count)
+
+        # assert event payload
+        mock_send_event.assert_called_with(
+            project_config,
+            mock_experiment,
+            mock_variation,
+            expected.flag_key,
+            expected.rule_key,
+            'feature-test',
+            expected.enabled,
+            'test_user',
+            {'browser': 'chrome'}
+        )
+
+    def test_decide__option__include_reasons__feature_test(self):
```

Comment:
These tests look good.
Can we add a few more tests validating "reasons" messages for:
- hit on everyone-else rule
- hit on none of the rules
- hit variation from "get_forced_variation"
- hit from "white_list"
- hit from user-profile-service
- any other important cases?